working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,937 @@
+try/catch statement, or in a type test `o is E` or a type cast `o as E`, or
+as the body of a type alias. It is also allowed to create a new instance
+where one or more extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+A compile-time error occurs if an extension type is used as a
+superinterface of a class or mixin, or if an extension type is used to
+derive a mixin.
+
+*So `class C extends E1 with E2 implements E3 {}` has three errors if `E1`,
+`E2`, and `E3` are extension types, and `mixin M on E1 implements E2 {}`
+has two errors.*
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>invokeExtensionMethod(Ext<S<sub>1</sub>, .. S<sub>k</sub>>, e).m(args)</code>
+and similarly for instance getters and operators. This rule also applies
+when a member access implicitly has the receiver `this`, and the static
+type of `this` is an extension type (*which can only occur in an extension
+type member declaration where the on-type is itself an extension type*).
+
+*That is, when the type of an expression is an extension type, all method
+invocations on that expression will invoke an extension method declared by
+that extension, and similarly for other member accesses. In particular, we
+cannot invoke an instance member when the receiver type is an extension
+type (unless the the extension type enables them explicitly, cf. the
+show/hide part specified in a later section).*
+
+Let `E` be an extension declaration named `Ext` with type parameters
+<code>X<sub>1</sub> extends B<sub>1</sub>, .. X<sub>k</sub> extends B<sub>k</sub></code>
+and on-type clause `on T`. Then we say that the _declared on-type_ of `Ext`
+is `T`, and the _instantiated on-type_ of
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>
+is
+<code>[S<sub>1</sub>/X<sub>1</sub>, .. S<sub>k</sub>/X<sub>k</sub>]T</code>.
+We will omit 'declared' and 'instantiated' from the phrase when it is clear
+from the context whether we are talking about the extension itself or a
+particular instantiation of a generic extension. For non-generic
+extensions, the on-type is the same in either case.
+
+An extension type `E` of the form
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>
+is a subtype of `Object?`, and a proper supertype of the
+instantiated on-type of `E`. In the case where the instantiated on-type of
+`E` is not a top type, `E` is also a proper subtype of `Object?`.
+
+*That is, the underlying on-type can only be recovered by an explicit cast
```

Comment:
explicit or implicit cast.

Reply:
I don't see what would be wrong with this. When the on-type is a top type it is always an upcast, so there is no requirement to make it explicit; that case is covered in line 390.

When the on-type is not a top type we have a proper subtype relationship `on-type <: E`, and the downcast must be made explicit.

Do we have a case where that is not true?

Reply:
```dart
To implicitDownCast<From, To extends From>(From x) {
  var l = <To>[];
  List<From> castThrough = l;
  castThrough.add(x);
  return l[0];
}

void main() {
  num x = 3;
  print(implicitDownCast<num, int>(x));
  x = 3.3;
  print(implicitDownCast<num, int>(x));
}
```